qiime2/core/type/signature.py

```diff
@@ -289,16 +289,36 @@ def decode_parameters(self, **kwargs):
 
     def check_types(self, **kwargs):
         for name, spec in self.signature_order.items():
-            if kwargs[name] not in spec.qiime_type:
-                # A type mismatch is unacceptable unless the value is None
-                # and this parameter's default value is None.
-                if not (spec.has_default() and
-                        spec.default is None and
-                        kwargs[name] is None):
+            # A type mismatch is unacceptable unless the value is None
+            # and this parameter's default value is None.
+            if ((kwargs[name] not in spec.qiime_type) and
+                    not (spec.has_default() and spec.default is None
+                         and kwargs[name] is None)):
+
+                # handle visualization passed as parameter
+                if isinstance(kwargs[name], qiime2.sdk.Visualization):
                     raise TypeError(
-                        "Parameter %r received an argument of type %r. An "
-                        "argument of subtype %r is required." % (
-                            name, kwargs[name].type, spec.qiime_type))
+                        "Parameter %r received a Visualization (*.qzv) as an "
```

Comment:
I like this error message! One small note though - not all `Visualization`s are qzv files --- if you are an Artifact API user you can generate a `Visualization` and "view" it in a notebook without ever needing to save it as a qzv.

Reply:
🗸

Reply:
I think we overcorrected in the wrong direction here --- for all the same reasons why we might not have "qzv files", we also can't guarantee that we have "objects" here, either.